drivers/hmis_supplemental/app/models/hmis_supplemental/field.rb

```diff
@@ -0,0 +1,114 @@
+###
+# Copyright 2016 - 2023 Green River Data Analysis, LLC
+#
+# License detail: https://github.com/greenriver/hmis-warehouse/blob/production/LICENSE.md
+###
+
+module HmisSupplemental
+  Field = Struct.new(:key, :label, :type, :multi_valued, :data_set, keyword_init: true) do
+    DELIMITER = '|'.freeze
+    TYPES = [
+      STRING_TYPE = 'string'.freeze,
+      ID_TYPE = 'id'.freeze,
+      INT_TYPE = 'int'.freeze,
+      FLOAT_TYPE = 'float'.freeze,
+      BOOLEAN_TYPE = 'boolean'.freeze,
+      DATE_TYPE = 'date'.freeze,
+    ].freeze
+    CLIENT_OWNER = 'client'.freeze
+    CLIENT_ID_FIELD = 'personal_id'.freeze
+    ENROLLMENT_OWNER = 'owner'.freeze
+    ENROLLMENT_ID_FIELD = 'enrollment_id'.freeze
+
+    def format_value(field_value)
+      value = field_value&.data
+      if multi_valued
+        return [] unless value.present?
+
+        return value.map { |i| format_single_value(i) }.join(', ')
+      end
+      format_single_value(value)
+    end
+
+    # @return [String, nil]
+    def format_single_value(value)
+      return if value.nil?
+
+      case type
+      when STRING_TYPE, ID_TYPE
+        value.to_s
+      when INT_TYPE
+        value.to_i.to_s
+      when FLOAT_TYPE
+        # 2 decimals seems like reasonable default. Could be configurable
+        format('%.2f', value.to_f)
+      when BOOLEAN_TYPE
+        value ? 'true' : 'false'
+      when DATE_TYPE
+        parse_date(value)&.to_fs
+      else
+        raise "unknown type: #{type}"
+      end
+    end
+
+    def row_owner_key(row)
+      owner_id = row_owner_id(row)
+      return nil unless owner_id
+
+      "#{owner_type}/#{row_owner_id(row)}"
```

Comment:
do we need to include `data_source_id` here?  neither PersonalID nor EnrollmentID are guaranteed to be unique across data sources.

Reply:
We could include it but the data source is on data_set already so it's not strictly necessary. The values don't really make sense without the field config which is on data set. FieldValue.for_owner(client/enrollment) joins to dataset to filter them by data source